src/machine/board_rpi3.go

```diff
@@ -0,0 +1,182 @@
+	reserved01               uint32 //0x24
+	OutputClear0             uint32 //0x28
+	OutputClear1             uint32 //0x2C
+	reserved03               uint32 //0x30
+	Level0                   uint32 //0x34
+	Level1                   uint32 //0x38
+	reserved04               uint32 //0x3C
+	EventDetectStatus0       uint32 //0x40
+	EventDetectStatus1       uint32 //0x44
+	reserved05               uint32 //0x48
+	RisingEdgeDetectEnable0  uint32 //0x4C
+	RisingEdgeDetectEnable1  uint32 //0x50
+	reserved06               uint32 //0x54
+	FallingEdgeDetectEnable0 uint32 //0x58
+	FallingEdgeDetectEnable1 uint32 //0x5C
+	reserved07               uint32 //0x60
+	HighDetectEnable9        uint32 //0x64
+	HighDetectEnable1        uint32 //0x68
+	reserved08               uint32 //0x6C
+	LowDetectEnable0         uint32 //0x70
+	LowDetectEnable1         uint32 //0x74
+	reserved09               uint32 //0x78
+	AsyncRisingEdgeDetect0   uint32 //0x7C
+	AsyncRisingEdgeDetect1   uint32 //0x80
+	reserved0A               uint32 //0x84
+	AsyncFallingEdgeDetect0  uint32 //0x88
+	AsyncFallingEdgeDetect1  uint32 //0x8C
+	reserved0B               uint32 //0x90
+	PullUpDownEnable         uint32 // 0x94
+	PullUpDownEnableClock0   uint32 //0x98
+	PullUpDownEnableClock1   uint32 //0x9C
+	reserved0C               uint32 //0xA0
+	test                     uint32 //0xA4
+}
+
+type SysTimerRegisterMap struct {
+	ControlStatus uint32 //0x00
+	CounterLower32 uint32 //0x04
+	CounterHigher32 uint32 //0x08
+	reservedGPU0 uint32 //0x0C
+	Compare1 uint32 //0x10
+	reservedGPU2 uint32 //0x14
+	Compare3 uint32 //0x18
+}
+
+func UARTInit() {
+	var r uint32
+
+	Aux.Enables |= 1  //enable UART1, AUX Mini uart
+	Aux.MiniUARTExtraControl=0
+	Aux.MiniUARTLineControl=3 //8 bits
+	Aux.MiniUARTModemControl=0
+	Aux.MiniUARTInterruptEnable=0
+	Aux.MiniUARTInterruptIdentify=0xC6 // disable interrupts
+	Aux.MiniUARTBAUD=270 // 115200 baud
+
+	// map UART1 to GPIO pins
+	r = GPIO.FuncSelect1
+	c := uint32((7 << 12) | (7 << 15)) // gpio14, gpio15
+	r &= ^c
+	r |= (2 << 12) | (2 << 15) // alt5
+	GPIO.FuncSelect1=r
+	GPIO.PullUpDownEnable=0
+
+	//sleep 150 cycles
+	r = 150
+	for r > 0 {
+		r--
+		arm.Asm("nop")
+	}
+
+	GPIO.PullUpDownEnableClock0=(1<<14)|(1<<15)
+
+	//sleep 150 cycles
+	r = 150
+	for r > 0 {
+		r--
+		arm.Asm("nop")
+	}
+
+	GPIO.PullUpDownEnableClock0=0 //flush gpio setup
+	Aux.MiniUARTExtraControl=3 //enable tx,rx
+}
+
+func UARTPutc(c uint8) {
+	// wait until we can send
+	for {
+		lineStatus:=Aux.MiniUARTLineStatus
+		if lineStatus&0x20 != 0 {
+			break
+		}
+		arm.Asm("nop")
+	}
+
+	// write the character to the buffer
+	c32 := uint32(c)
+	Aux.MiniUARTData=c32
+}
+
+func UARTGetc() uint8 {
+	for {
+		lineStatus := Aux.MiniUARTLineStatus
+		if lineStatus&0x01 != 0 {
+			break
+		}
+		arm.Asm("nop")
+	}
+	r:=Aux.MiniUARTData
+	return uint8(r)
+}
+
+func UARTPuts(s string) {
+	for i := 0; i < len(s); i++ {
+		UARTPutc(s[i])
+	}
+}
+
+func Abort() {
+	arm.Asm("nop")
```

Comment:
How does this function as an abort? Wouldn't you need a loop around this?

Reply:
Changed to wfe (wait for event) which means "go to sleep and wait for something" ... that will never happen, in this case.

Reply:
I would still recommend a loop around the `wfe` here. Even though it _should_ not continue after the abort, it may happen due to unexpected circumstances. A loop will only cost one extra instruction.

Also, all other targets put `abort` in the runtime and keep it unexported. There is no reason a program should call it directly: if programs need a way to exit then `os.Exit` should be implemented instead (which still doesn't make a lot of sense on an embedded system).